ecpy_hqc_legacy/tasks/tasks/instr/spdev_tasks.py

```diff
@@ -98,56 +117,145 @@ def perform(self):
         """
         if self.driver.owner != self.name:
             self.driver.owner = self.name
-
             self.driver.configure_board()
 
+        avg_bef_demod = True if self.average == 'Avg before demod' else False
+
         records_number = self.format_and_eval_string(self.records_number)
         delay = self.format_and_eval_string(self.delay)*1e-9
         duration = self.format_and_eval_string(self.duration)*1e-9
+        sampling_rate = self.format_and_eval_string(self.sampling_rate)
 
         channels = (self.ch1_enabled, self.ch2_enabled)
-        ch1, ch2 = self.driver.get_traces(channels, duration, delay,
-                                          records_number)
+
+        ch = self.driver.get_traces(channels, duration, delay,
+                                    records_number, average=avg_bef_demod)
+
+        average = False if self.average == 'No avg' else True
+
+        ch1 = ch[0]
+        ch2 = ch[1]
+
         if self.ch1_enabled:
             f1 = self.format_and_eval_string(self.freq_1)*1e6
-            phi1 = np.linspace(0, 2*np.pi*f1*duration, len(ch1))
+
+            # Remove point that do not belong to a full period.
+            samples_per_period = int(sampling_rate/f1)
```

Comment:
Be careful here that int will truncate the result of the division. using int(round()) is usually safer.